plugins/gateway/tyk/v3.2.2/plugin.go

```diff
@@ -73,27 +88,38 @@ func PostGetAPIDefinition(_ http.ResponseWriter, r *http.Request) {
 		logger.Error("Failed to get api definition")
 		return
 	}
-	// set the apiDefinition since we dont get it in the response phase
-	ctx.SetDefinition(r, apiDefinition)
-
-	requestTime := time.Now().UTC().UnixNano() / int64(time.Millisecond) // UnixMilli supported only from go 1.17
 
 	session := ctx.GetSession(r)
-	if session == nil {
-		session = &user.SessionState{MetaData: map[string]interface{}{common.RequestTimeContextKey: requestTime}}
-	} else if session.MetaData == nil {
-		session.MetaData = map[string]interface{}{common.RequestTimeContextKey: requestTime}
-	} else {
-		session.MetaData[common.RequestTimeContextKey] = requestTime
+	if traceSamplingEnabled && TraceSamplingManager != nil {
+		host, _ := getHostAndPortFromTargetURL(apiDefinition.Proxy.TargetURL)
+		session = setIgnoreHostOnSession(session, host)
 	}
-	// set request time on session metadata
+
+	session = setRequestTimeOnSession(session)
+
 	ctx.SetSession(r, session, false, false)
+
+	// set the apiDefinition since we dont get it in the response phase
+	ctx.SetDefinition(r, apiDefinition)
 }
 
 // Called during response phase.
 //nolint:deadcode
 func ResponseSendTelemetry(_ http.ResponseWriter, res *http.Response, req *http.Request) {
 	logger.Info("Handling telemetry")
+
+	if traceSamplingEnabled {
+		metadata := ctx.GetSession(req).MetaData
+		ignoreHost, ok := metadata[common.IgnoreHost].(string)
```

Comment:
can't we run `TraceSamplingManager.ShouldTrace(host)` here and avoid writing/reading to session?

Reply:
yes. thank you.